src/pmdas/farm/pmda.c

```diff
@@ -0,0 +1,1907 @@
+/*
+ * Seagate Field Accessible Reliability Metrics (FARM) Log
+ *
+ * Copyright (c) 2023 Red Hat.
+ * 
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+ 
+ #include "pmapi.h"
+ #include "pmda.h"
+ #include "domain.h"
+ 
+ #include "pmdafarm.h"
+ 
+ static int _isDSO = 1; /* for local contexts */
+ static char *farm_setup_lsblk;
+ static char *farm_setup_smartctl;
+ 
+ pmdaIndom indomtable[] = {
```

Comment:
This PMDA will benefit from using additional indoms I think.  It looks like you have a 'compound indom' situation here - we tend to solve this using '::' separator on the external instance names, and use of the pmdaCache APIs.  Have a look at the cgroup.io.stat.rbytes metric as an example, which combines cgroups and devices, e.g.

```
pminfo -f cgroup.io.stat.rbytes
    inst [...]
    inst [34 or "/system.slice::vda"] value 75020690432
```

These metrics also have labels for querying the individual parts of the instance, e.g.

```
pminfo -l cgroup.io.stat.rbytes
    [...]
    inst [34 or "/system.slice::vda"] labels {"agent":"proc","cgroup":"/system.slice","device_name":"vda","device_type":"block","domainname":"...","groupid":1000,"hostname":"fedora","machineid":"...","userid":1000}
```